### PR TITLE
Feature/order management

### PR DESCRIPTION
**Pull request by HaiderZaman77**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -38,6 +45,10 @@ group :development, :test do
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
 end
 
+#cancancan gem
+gem 'cancancan'
+
+
```
Comment (abbasraza172) [inline Gemfile]: extra line added here. please remove

File: Gemfile [hunk 2]
```diff
@@ -24,7 +24,14 @@ gem 'jbuilder', '~> 2.7'
...
 
 #gem font-awesome
+#gem "font-awesome-rails"
+
+#devise
+gem 'devise'
+
+
 gem "font-awesome-rails"
+gem 'jquery-rails'
```
Comment (abbasraza172) [inline Gemfile]: extra line added here. please remove

File: app/assets/stylesheets/application.css [hunk 3]
```diff
@@ -12,4 +12,4 @@
  *
  *= require_tree
  *= require_self
- */
+ */
```
Comment (abbasraza172) [inline app/assets/stylesheets/application.css]: please add new line at end.

File: app/assets/stylesheets/bootstrap.css [hunk 4]
```diff
@@ -6162,4 +6162,4 @@ a.badge:focus {
 
 .affix {
   position: fixed;
-}
+}
```
Comment (abbasraza172) [inline app/assets/stylesheets/bootstrap.css]: please add new line at end.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,5 +1,12 @@
 class ApplicationController < ActionController::Base
     before_action :set_subdomain
+    before_action :configure_permitted_parameters, if: :devise_controller?
+
+
+    def configure_permitted_parameters
+        # devise_parameter_sanitizer.for(:sign_up) { |u| u.permit(:name, :email, :password, :password_confirmation) }
```
Comment (abbasraza172) [inline app/controllers/application_controller.rb]: please remove commented code.

File: app/controllers/seller_products_controller.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        redirect_to new_user_session_path
+      end
+    end
+    
+    def new
+      @product = Product.new
+    end
+    
+    def create
+      # product_params[:category_ids] = product_params[:category_ids].reject{|c| c.blank? }
```
Comment (abbasraza172) [inline app/controllers/seller_products_controller.rb:18]: please remove commented code.

File: app/javascript/packs/seller.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+
+require('jquery')
+require("packs/jquery-1.10.2")
+require("packs/jquery.metisMenu")
+require("packs/bootstrap-seller")
+require("packs/custom")
+require("packs/dropdownmin")
+require("packs/dropdownjquery")
+// window.ddSelect =   require("packs/dropdownselect")
```
Comment (abbasraza172) [inline app/javascript/packs/seller.js]: please remove commented code.

File: app/models/ability.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    # Define abilities for the passed in user here. For example:
+    #
+    if user.seller?  # additional permissions for logged in users (they can read their own posts)
```
Comment (abbasraza172) [inline app/models/ability.rb]: is the comment correct. post ?

File: app/models/ability.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      can :manage, Product , business_id: user.business_id
+      can :update, Order , user_id: user.id
+    end
+
+    if user.admin?  # additional permissions for logged in users (they can read their own posts)
```
Comment (abbasraza172) [inline app/models/ability.rb]: same here.

File: app/models/product_category.rb [hunk 10]
```diff
@@ -4,6 +4,6 @@ class ProductCategory < ApplicationRecord
   belongs_to :category
 
   #validations
-  validates :product_id, presence: true
+  #validates :product_id, presence: true
```
Comment (abbasraza172) [inline app/models/product_category.rb]: please remove the commented code from here.

File: app/models/user.rb [hunk 11]
```diff
@@ -26,4 +35,12 @@ class User < ApplicationRecord
   def has_access(right)
     User::ACCESS_RIGHTS[self.role].include?(right)
   end
+
+  def create_business
```
Comment (abbasraza172) [inline app/models/user.rb:41]: please add code docs. here

File: app/views/categories/_form.html.erb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    <%= form.radio_button :status, "True" %>
+    <%= form.label :status_true, "True" %>
+    <%= form.radio_button :status, "False" %>
+    <%= form.label :status_false, "False" %>
+
+  <div class="actions">
+    <%= form.submit  class: 'btn btn-warning mb-5' %>
+    <%= link_to 'Back', categories_path , class: 'btn btn-primary mt-3'%>
+  </div>
+<% end %>
```
Comment (abbasraza172) [inline app/views/categories/_form.html.erb]: add new line here.

File: app/views/categories/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<h1 class="bg-info text-center">Update Category</h1>
+<%= render 'form', category: @category %>
+<%= link_to 'Back', categories_path , class: 'btn btn-primary mt-3'%>
```
Comment (abbasraza172) [inline app/views/categories/edit.html.erb]: add new line here.

File: app/views/categories/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+                        <%end%>
+                        <% if can? :edit, category %>
+                            <%= link_to 'Edit Category', edit_category_path(category) , class: 'btn btn-sm btn-info' %>
+                        <%end%>
+                    </div>
+                </div>
+            </li>    
+        <% end %>       
+    </ul>
+</div>
```
Comment (abbasraza172) [inline app/views/categories/index.html.erb]: add new line here.

File: app/views/categories/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,9 @@
+<div class="container">
+    <% if can? :new, @category %>
+        <h1 class="bg-info text-center">New Category</h1>
+        <%= render 'shared/cerrors' %>
+        <%= render 'form', category: @category %>
+    <%else%>
+    <h1 class="container">You Don't have the permissions for this action</h1>
+    <%end%>
+</div>
```
Comment (abbasraza172) [inline app/views/categories/new.html.erb]: add new line here.

File: app/views/orders/edit.html.erb [hunk 16]
```diff
@@ -0,0 +1,5 @@
+<div class="container">
+    <h1 class="bg-info text-center">update order Status</h1>
+    <%= render 'form', order: @order %>
+    <%= link_to 'Back', order_path , class: 'btn btn-primary mt-3'%>
+</div>
```
Comment (abbasraza172) [inline app/views/orders/edit.html.erb]: add new line here.

File: app/views/shared/_notification-bar.html.erb [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+                </span>
+                </a>
+                <a href="#" class="list-group-item">
+                    <i class="fa fa-envelope fa-fw"></i>Feedback Received
+                <span class="pull-right text-muted small"><em>27 minutes ago</em>
+                </span>
+                </a>
+                <a href="#" class="list-group-item">
+                    <i class="fa fa-tasks fa-fw"></i><%=Order.where(status: "pending").count%> Pending Orders
+                <%# <span class="pull-right text-muted small"><em><%= distance_of_time_in_words(Order.where(status: "pending").created_at, Time.now, include_seconds: true) %>
```
Comment (abbasraza172) [inline app/views/shared/_notification-bar.html.erb]: please remove the commented code.

File: app/views/shared/_notification-bar.html.erb [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+                </span>
+                </a>
+                <a href="#" class="list-group-item">
+                    <i class="fa fa-warning fa-fw"></i><%=Order.where(status: "failed").count%> Rejected Orders
+                <%# <span class="pull-right text-muted small"><em><%= distance_of_time_in_words(Order.where(status: "pending").created_at, Time.now, include_seconds: true) %>
```
Comment (abbasraza172) [inline app/views/shared/_notification-bar.html.erb]: please remove the commented code.

File: app/views/shared/_notification-bar.html.erb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+                </span>
+                </a>
+                <a href="#" class="list-group-item">
+                    <i class="fa fa-bolt fa-fw"></i><%=Order.where(status: "canceled").count%> Cancelled Orders
+                <%# <span class="pull-right text-muted small"><em><%= distance_of_time_in_words(Order.where(status: "pending").created_at, Time.now, include_seconds: true) %>
```
Comment (abbasraza172) [inline app/views/shared/_notification-bar.html.erb]: please remove the commented code.

File: app/views/shared/_notification-bar.html.erb [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+                </span>
+                </a>
+                <a href="#" class="list-group-item">
+                    <i class="fa fa-shopping-cart fa-fw"></i><%=Order.where(status: "pending").count%> New Order Placed
+                <%# <span class="pull-right text-muted small"><em><%= distance_of_time_in_words(Order.where(status: "pending").last.created_at, Time.now, include_seconds: true) %>
```
Comment (abbasraza172) [inline app/views/shared/_notification-bar.html.erb]: please remove the commented code.

File: app/views/shared/_review-crr.html.erb [hunk 21]
```diff
@@ -0,0 +1,45 @@
+<div id="reviews" class="carousel slide" data-ride="carousel">
+
+    <div class="carousel-inner">
+        <div class="item active">
+
+            <div class="col-md-10 col-md-offset-1">
+
+                <h4><i class="fa fa-quote-left"></i>Lorem ipsum dolor sit amet, consectetur adipiscing Lorem ipsum dolor sit amet, consectetur adipiscing elit onec molestie non sem vel condimentum. <i class="fa fa-quote-right"></i></h4>
+                <div class="user-img pull-right">
+                    <%# <img src="assets/img/user.gif" alt="" class="img-u image-responsive" /> %>
```
Comment (abbasraza172) [inline app/views/shared/_review-crr.html.erb]: please remove the commented code.

File: app/views/shared/_review-crr.html.erb [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+                </div>
+                <h5 class="pull-right"><strong class="c-black">Lorem Dolor</strong></h5>
+            </div>
+        </div>
+        <div class="item">
+            <div class="col-md-10 col-md-offset-1">
+
+                <h4><i class="fa fa-quote-left"></i>Lorem ipsum dolor sit amet, consectetur adipiscing Lorem ipsum dolor sit amet, consectetur adipiscing elit onec molestie non sem vel condimentum. <i class="fa fa-quote-right"></i></h4>
+                <div class="user-img pull-right">
+                    <%# <img src="assets/img/user.png" alt="" class="img-u image-responsive" /> %>
```
Comment (abbasraza172) [inline app/views/shared/_review-crr.html.erb]: please remove the commented code.

File: app/views/shared/_review-crr.html.erb [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+                <div class="user-img pull-right">
+                    <%# <img src="assets/img/user.gif" alt="" class="img-u image-responsive" /> %>
+                </div>
+                <h5 class="pull-right"><strong class="c-black">Lorem Dolor</strong></h5>
+            </div>
+        </div>
+        <div class="item">
+            <div class="col-md-10 col-md-offset-1">
+
+                <h4><i class="fa fa-quote-left"></i>Lorem ipsum dolor sit amet, consectetur adipiscing Lorem ipsum dolor sit amet, consectetur adipiscing elit onec molestie non sem vel condimentum. <i class="fa fa-quote-right"></i></h4>
```
Comment (abbasraza172) [inline app/views/shared/_review-crr.html.erb:17]: what is this.

File: app/views/shared/_review-crr.html.erb [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+                <div class="user-img pull-right">
+                    <%# <img src="assets/img/user.png" alt="" class="img-u image-responsive" /> %>
+                </div>
+                <h5 class="pull-right"><strong class="c-black">Lorem Dolor</strong></h5>
```
Comment (abbasraza172) [inline app/views/shared/_review-crr.html.erb:20]: that is this.

File: app/views/shared/_review-crr.html.erb [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+            </div>
+
+        </div>
+        <div class="item">
+            <div class="col-md-10 col-md-offset-1">
+
+                <h4><i class="fa fa-quote-left"></i>Lorem ipsum dolor sit amet, consectetur adipiscing  Lorem ipsum dolor sit amet, consectetur adipiscing elit onec molestie non sem vel condimentum. <i class="fa fa-quote-right"></i></h4>
```
Comment (abbasraza172) [inline app/views/shared/_review-crr.html.erb:27]: what is this

File: app/views/shared/_review-crr.html.erb [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+                <div class="user-img pull-right">
+                    <%# <img src="assets/img/user.gif" alt="" class="img-u image-responsive" /> %>
```
Comment (abbasraza172) [inline app/views/shared/_review-crr.html.erb]: please remove the commented code.

File: app/views/shared/_side-menu.html.erb [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+                </div>
+
+            </li>
+
+
+            <li>
+                <%= link_to 'Dashboard', seller_path , class: 'fa fa-dashboard active-menu'%>
+            </li>
+            <%# <li>
+                <a href="#"><i class="fa fa-desktop "></i>Product Categories<span class="fa arrow"></span></a>
```
Comment (abbasraza172) [inline app/views/shared/_side-menu.html.erb]: please remove the commented code.

File: config/database.yml [hunk 28]
```diff
@@ -14,7 +14,7 @@ default: &default
   encoding: utf8mb4
```
Comment (abbasraza172) [inline config/database.yml:14]: we dont add this file in commit.

File: app/views/shared/_review-crr.html.erb (hunk 23 above)
Comment (HaiderZaman77) [reply app/views/shared/_review-crr.html.erb:17]: Sample carousel that will be populated with some comments by customer

**Review by abbasraza172 — APPROVED** [review]

**Pull request merged**